src/openrct2-ui/windows/Viewport.cpp

```diff
@@ -10,192 +10,220 @@
+static constexpr const int32_t WW = 200;
+static constexpr const int32_t WH = 200;
+
+static constexpr ScreenSize VIEWPORT_BUTTON = {24, 24};
+
+#pragma endregion
+
+static rct_widget window_viewport_widgets[] =
+{
     WINDOW_SHIM(WINDOW_TITLE, WW, WH),
     MakeWidget({      0, 14}, { WW - 1, WH - 1}, WindowWidgetType::Resize,   WindowColour::Secondary                                         ), // resize
     MakeWidget({      3, 17}, {WW - 26, WH - 3}, WindowWidgetType::Viewport, WindowColour::Primary                                           ), // viewport
-    MakeWidget({WW - 25, 17}, {     24,     24}, WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_G2_ZOOM_IN,  STR_ZOOM_IN_TIP       ), // zoom in
-    MakeWidget({WW - 25, 41}, {     24,     24}, WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_G2_ZOOM_OUT, STR_ZOOM_OUT_TIP      ), // zoom out
-    MakeWidget({WW - 25, 65}, {     24,     24}, WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_LOCATE,      STR_LOCATE_SUBJECT_TIP), // locate
+    MakeWidget({WW - 25, 17}, VIEWPORT_BUTTON,   WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_G2_ZOOM_IN,  STR_ZOOM_IN_TIP       ), // zoom in
+    MakeWidget({WW - 25, 41}, VIEWPORT_BUTTON,   WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_G2_ZOOM_OUT, STR_ZOOM_OUT_TIP      ), // zoom out
+    MakeWidget({WW - 25, 65}, VIEWPORT_BUTTON,   WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_LOCATE,      STR_LOCATE_SUBJECT_TIP), // locate
     { WIDGETS_END },
 };
 
-static void window_viewport_mouseup(rct_window *w, rct_widgetindex widgetIndex);
-static void window_viewport_resize(rct_window *w);
-static void window_viewport_update(rct_window *w);
-static void window_viewport_invalidate(rct_window *w);
-static void window_viewport_paint(rct_window *w, rct_drawpixelinfo *dpi);
-
-static rct_window_event_list window_viewport_events([](auto& events)
-{
-    events.mouse_up = &window_viewport_mouseup;
-    events.resize = &window_viewport_resize;
-    events.update = &window_viewport_update;
-    events.invalidate = &window_viewport_invalidate;
-    events.paint = &window_viewport_paint;
-});
 // clang-format on
 
-static int32_t _viewportNumber = 1;
-
-/**
- * Creates a custom viewport window.
- */
-rct_window* window_viewport_open()
+class ViewportWindow final : public Window
 {
-    rct_window* w = WindowCreateAutoPos(WW, WH, &window_viewport_events, WC_VIEWPORT, WF_RESIZABLE);
-    w->widgets = window_viewport_widgets;
-    w->enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
-    w->number = _viewportNumber++;
-
-    // Create viewport
-    viewport_create(w, w->windowPos, w->width, w->height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
-    rct_window* mainWindow = window_get_main();
-    if (mainWindow != nullptr)
+private:
+    int32_t _viewportWindowNumber = 1;
```

Comment:
I realise this was in the original window but what is the point of this variable. It's identical to the window number afaik and therefore there is no point in it @IntelOrca Could you confirm what the idea with it was. All the vehicle, peep, banner windows just use the window number to be unique.

Reply:
I only created this var so I could increment it until I found the highest unused number. I can remove this var and instead work directly with the window->number var.

Reply:
Just seemed cleaner (to me) to only touch the window->number once, ie. after I'd found what to set it to.
Probably doesn't matter though.

Reply:
Theres is no difference between the two variables access wise

Reply:
Ahh you don't need it you can just pass by reference into the lambda.
```
window_visit_each([&](rct_window* w) { GetFreeViewportNumber(w, number); });
```
Then just modify get free viewport number to take the number by reference.

Reply:
I think it was so that it displayed the number in the title of the window, and also so each viewport window had a unique window number as they all have identical class.

Reply:
I've removed the _viewportWindowNumber variable and just use the window->number to build the title of the window.